File: word-doc-problems/binary-tree-problems/problem3.py
# // Given an array of numbers sorted in ascending order, find the ceiling of a given number ‘key’. The ceiling of the ‘key’ will be the smallest element in the given array greater than or equal to the ‘key’.
# // for example:
# // Input: [4, 6, 10], key=6
# // Output: 1
# // explanation: the smallest number greater than or equal to '6' is 6 having index 1

# Input: [1, 3, 8, 10, 15], key = 12
# Output: 4
# explanation: the smallest number greater than or equal to 12 is 15 having index 4


def search_ceiling(arr, key):
  n = len(arr)
  if key > arr[n - 1]:
    return -1

  start, end = 0, n -1
  while start <= end:
    mid= start + (end - start) // 2
    if key < arr[mid]:
      end = mid - 1
    elif key > arr[mid]:
      start = mid + 1
    else:
      return mid
  return start

File: word-doc-problems/array-problems/problem3.js
// Given an array a of n numbers and a count k
// find the k largest values in the array a.
// Example: a=[5, 1, 3, 6, 8, 2, 4, 7], k=3  ⇒  [6, 8, 7]

function largestValues(a, k) {
  const sorted = a.sort((x, y) => x - y);
  return sorted.reduce((initial, current) => {
    const index = [];
    for (let i = 0; i < k; i++) {
      if (current === sorted[sorted.length - i - 1]) {
        index.push(current);
      }
    }
    return initial.concat(index);
  }, []);
}
const a = [5, 1, 3, 6, 8, 2, 4, 7];
const k = 3;
console.log(largestValues(a, k));